File: tests/src/test_roundtrip.cpp
#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>
#include <armadillo>
#include <carma>

namespace py = pybind11;

namespace carma {
namespace tests {

py::array_t<double> test_mat_roundtrip(py::array_t<double>& arr) {
    // call function to be tested
    arma::Mat<double> M = carma::arr_to_mat<double>(std::move(arr));
    return carma::mat_to_arr(M);
}

py::array_t<double> test_row_roundtrip(py::array_t<double>& arr) {
    // call function to be tested
    arma::Row<double> M = carma::arr_to_row<double>(std::move(arr));
    return carma::row_to_arr(M);
}

py::array_t<double> test_col_roundtrip(py::array_t<double>& arr) {
    // call function to be tested
    arma::Col<double> M = carma::arr_to_col<double>(std::move(arr));
    return carma::col_to_arr(M);
}

py::array_t<double> test_cube_roundtrip(py::array_t<double>& arr) {
    // call function to be tested
    arma::Cube<double> M = carma::arr_to_cube<double>(std::move(arr));
    return carma::cube_to_arr(M);
}

}  // namespace tests
}  // namespace carma

void bind_test_mat_roundtrip(py::module& m) {
    m.def("mat_roundtrip", &carma::tests::test_mat_roundtrip, "Test mat_roundtrip");
}

void bind_test_row_roundtrip(py::module& m) {
    m.def("row_roundtrip", &carma::tests::test_row_roundtrip, "Test row_roundtrip");
}

void bind_test_col_roundtrip(py::module& m) {
    m.def("col_roundtrip", &carma::tests::test_col_roundtrip, "Test col_roundtrip");
}

void bind_test_cube_roundtrip(py::module& m) {
    m.def("cube_roundtrip", &carma::tests::test_cube_roundtrip, "Test cube_roundtrip");
}

File: tests/test_roundtrip.py
"""Test numpy array to arma and back conversion functions."""
import numpy as np

import test_carma as carma


def test_mat_roundtrip():
    """Test mat_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(50, 2)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.mat_roundtrip(sample)
    assert np.allclose(og_sample, out)

def test_mat_roundtrip_large():
    """Test mat_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(1000, 1000)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.mat_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_mat_roundtrip_small():
    """Test mat_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(3, 3)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.mat_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_mat_roundtrip_c_order():
    """Test mat_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(50, 2)), dtype=np.float64, order='C'
    )
    sample = og_sample.copy()
    out = carma.mat_roundtrip(sample)
    assert np.allclose(og_sample, out)

def test_mat_roundtrip_large_c_order():
    """Test mat_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(1000, 1000)), dtype=np.float64, order='C'
    )
    sample = og_sample.copy()
    out = carma.mat_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_mat_roundtrip_small_c_order():
    """Test mat_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(3, 3)), dtype=np.float64, order='C'
    )
    sample = og_sample.copy()
    out = carma.mat_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_row_roundtrip():
    """Test row_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(50)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.row_roundtrip(sample)
    assert np.allclose(og_sample, out)

def test_row_roundtrip_large():
    """Test row_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(1000)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.row_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_row_roundtrip_small():
    """Test row_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(3)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.row_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_col_roundtrip():
    """Test col_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(50)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.col_roundtrip(sample)
    assert np.allclose(og_sample, out.ravel())

def test_col_roundtrip_large():
    """Test col_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(1000)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.col_roundtrip(sample)
    assert np.allclose(og_sample, out.ravel())


def test_col_roundtrip_small():
    """Test col_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(3)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.col_roundtrip(sample)
    assert np.allclose(og_sample, out.ravel())


def test_cube_roundtrip():
    """Test cube_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(50, 3, 2)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.cube_roundtrip(sample)
    assert np.allclose(og_sample, out)

def test_cube_roundtrip_large():
    """Test cube_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=(1000, 10, 2)), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.cube_roundtrip(sample)
    assert np.allclose(og_sample, out)


def test_cube_roundtrip_small():
    """Test cube_roundtrip."""
    og_sample = np.asarray(
        np.random.normal(size=((2, 2, 2))), dtype=np.float64, order='F'
    )
    sample = og_sample.copy()
    out = carma.cube_roundtrip(sample)
    assert np.allclose(og_sample, out)
